src/inbox/mod_inbox.erl

```diff
@@ -430,36 +449,39 @@ text_single_form_field(Var, DefaultValue) ->
 %%%%%%%%%%%%%%%%%%%
 %% iq-set
 -spec query_to_params(mongooseim:host_type(), QueryEl :: exml:element()) ->
-    get_inbox_params() | {error, bad_request, Msg :: binary()}.
+    get_inbox_params() | {error, bad_request, binary()}.
 query_to_params(HostType, QueryEl) ->
-    case form_to_params(HostType, exml_query:subelement_with_ns(QueryEl, ?NS_XDATA)) of
-        {error, bad_request, Msg} ->
-            {error, bad_request, Msg};
-        Params ->
-            case maybe_rsm(exml_query:subelement_with_ns(QueryEl, ?NS_RSM)) of
-                {error, Msg} -> {error, bad_request, Msg};
-                undefined -> Params;
-                Rsm -> Params#{limit => Rsm}
-            end
-    end.
-
--spec maybe_rsm(exml:element() | undefined) ->
-    undefined | non_neg_integer() | {error, binary()}.
-maybe_rsm(#xmlel{name = <<"set">>,
-                 children = [#xmlel{name = <<"max">>,
-                                    children = [#xmlcdata{content = Bin}]}]}) ->
-    case mod_inbox_utils:maybe_binary_to_positive_integer(Bin) of
-        {error, _} -> {error, wrong_rsm_message()};
-        0 -> undefined;
-        N -> N
+    Form = form_to_params(HostType, exml_query:subelement_with_ns(QueryEl, ?NS_XDATA)),
+    Rsm = jlib:rsm_decode(QueryEl),
+    build_params(Form, Rsm).
+
+-spec build_params(get_inbox_params() | {error, bad_request, binary()}, none | jlib:rsm_in()) ->
+    get_inbox_params() | {error, bad_request, binary()}.
+build_params({error, bad_request, Msg}, _) ->
+    {error, bad_request, Msg};
+build_params(_, #rsm_in{max = Max, index = Index}) when Max =:= error; Index =:= error ->
+    {error, bad_request, <<"bad-request">>};
+build_params(Params, none) ->
+    Params;
+build_params(Params, #rsm_in{max = Max, id = undefined}) when Max =/= undefined ->
+    Params#{limit => Max};
+build_params(Params0, Rsm) ->
+    Params = maps:without([start, 'end'], Params0),
+    build_params_with_rsm(Params#{rsm => Rsm}, Rsm).
+
+build_params_with_rsm(Params, #rsm_in{max = Max, id = <<>>, direction = before}) ->
+    Params#{limit => Max, order => asc, start => 0};
+build_params_with_rsm(Params, #rsm_in{max = Max, id = <<>>, direction = aft}) ->
+    Params#{limit => Max, 'end' => (1 bsl 63 - 1)};
```

Comment:
Why do we have to pass an integer here? If we skip it, wouldn't the DB query just omit this unnecessary guard parameter?